packages/cli/generation/ir-generator/src/utils/getCyclicTypes.ts

```diff
@@ -0,0 +1,203 @@
+import { RelativeFilePath } from "@fern-api/fs-utils";
+import {
+    isRawDiscriminatedUnionDefinition,
+    isRawObjectDefinition,
+    isRawUndiscriminatedUnionDefinition,
+    RawSchemas
+} from "@fern-api/yaml-schema";
+import { constructFernFileContext, FernFileContext } from "../FernFileContext";
+import { ResolvedType } from "../resolvers/ResolvedType";
+import { TypeResolver } from "../resolvers/TypeResolver";
+import {
+    CASINGS_GENERATOR,
+    getAllPropertiesForType,
+    ObjectPropertyWithPath,
+    TypeName,
+    TypePropertyArgs
+} from "./getAllPropertiesForObject";
+
+export declare type CyclicTypeArgs = TypePropertyArgs & {
+    childName: TypeName;
+    seen?: Set<TypeName>;
+};
+
+declare type TypeOrProperty = ResolvedTypeWithKeys | ObjectPropertyWithPath;
+
+declare type UnionValue =
+    | string
+    | {
+          docs?: string | undefined;
+          name?: string | undefined;
+          type?: string | Record<string, never> | undefined;
+          key?: string | { value: string; name?: string | undefined } | undefined;
+      };
+
+interface ResolvedTypeWithKeys {
+    keys: string[];
+    type: ResolvedType | null;
+}
+
+interface TypeParams {
+    keys: string[];
+    name: TypeName;
+    type: ResolvedType;
+    filePath: RelativeFilePath;
+}
+
+/**
+ * This function will return the first found path to a named child type from a given type. It will only search within the file of the given type.
```

Comment:
> It will only search within the file of the given type.

Seems like a big limitation, no?

Reply:
https://github.com/fern-api/fern/blob/main/packages/cli/yaml/validator/src/rules/no-duplicate-declarations/no-duplicate-declarations.ts#L15 Is an example of a rule that is able to check for name conflicts across files (admittedly much easier)

Reply:
> > It will only search within the file of the given type.
> 
> 
> 
> Seems like a big limitation, no? 

It's not necessary for this use-case, since the existing rule for cyclic imports would already prevent any cyclic types across multiple files. 

Since that narrows our search to only cycles defined within a single file, we can skip walking to any other files and allow the rule to catch them when the main visitor for the rules reaches them normally.